src/cljs/athens/db.cljs

```diff
@@ -525,44 +525,24 @@
 ;; history
 
 (defonce history (atom []))
-(def ^:const history-limit 10)
-
-
-(defn drop-tail
-  [xs pred]
-  (loop [acc []
-         xs  xs]
-    (let [x (first xs)]
-      (cond
-        (nil? x) acc
-        (pred x) (conj acc x)
-        :else  (recur (conj acc x) (next xs))))))
-
-
-(defn trim-head
-  [xs n]
-  (vec (drop (- (count xs) n) xs)))
-
-
-(defn find-prev
-  [xs pred]
-  (last (take-while #(not (pred %)) xs)))
-
-
-(defn find-next
-  [xs pred]
-  (fnext (drop-while #(not (pred %)) xs)))
-
+#_(def ^:const history-limit 10)
 
+;; this gives us customization options
+;; now if there is a pattern for a tx then the datoms can be
+;; easily modified(mind the order of datoms) to add a custom undo/redo strategy
+;; Not seeing a use case now, but there is an option to do it
 (d/listen! dsdb :history
            (fn [tx-report]
-             (let [{:keys [db-before db-after]} tx-report]
-               (when (and db-before db-after)
-                 (swap! history (fn [h]
-                                  (-> h
-                                      (drop-tail #(identical? % db-before))
-                                      (conj db-after)
-                                      (trim-head history-limit))))))))
+             (when-not (or (->> tx-report :tx-data (some (fn [datom]
+                                                           (= (nth datom 1)
+                                                              :from-undo-redo))))
+                           (->> tx-report :tx-data empty?))
+               (swap! history (fn [buff]
+                                (->> buff (remove (fn [[_ applied? _]]
+                                                    (not applied?))))))
+               (swap! history conj [(-> tx-report :tx-data first vec (nth 3))
```

Comment:
When I undo then redo, I get the following error with `nth`:

![image](https://user-images.githubusercontent.com/8952138/111023305-1b810600-838d-11eb-98b0-c787b061b1c6.png)

Maybe here?

Reply:
I will say it undo is a lot faster!

Reply:
Regarding the bug -- lists and lazy seq gotchas!

Regarding rtc -- the idea of db + tx1 + tx2 gives the final state is something that we can leverage. I won't say this is the final design idea but this is a good naive starting point.  (there is no restriction tx1 and tx2 should come from the same source 😉)
Challenge in rtc will be understanding n states(for n users) and coming to a conflict-free state on the server-side and pushing that state in a sane manner to all users(show errors or copy old blocks when there is a conflict between server data and user data -- since the log is append-only we are bound to run into this)